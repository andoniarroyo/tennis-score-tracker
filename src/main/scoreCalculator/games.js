// @flow
import type { UpdateSideEffectType, ScoreContextType } from '../matchScoreCalculator';

const MINIMUM_GAMES_TO_WIN_A_SET = 6;
const MINIMUM_GAMES_DIFFERENCE_TO_WIN_A_SET = 2;

export type NextGameSideEffectType = {
    isFinishingSet: boolean,
    updates: Array<UpdateSideEffectType>,
}

/**
 * Calculates the new games value for the winner of the point
 * @param {number} winnerGames Won games by the winner
 * @param {number} loserGames Won games by the loser
 * @return {number} The new won games by the winner of the point
 * @private
 */
const calculateWinnerGames = (winnerGames: number, loserGames: number): number => {
  const newWinnerGames = winnerGames + 1;
  return (newWinnerGames >= MINIMUM_GAMES_TO_WIN_A_SET &&
    newWinnerGames - loserGames >= MINIMUM_GAMES_DIFFERENCE_TO_WIN_A_SET) ? 0 : (newWinnerGames);
};

/**
 * Calculates the new games value for the loser of the point
 * @param {number} newWinnerGames Won games by the winner (including the current point)
 * @param {number} loserGames Won games by the loser
 * @return {number} The new won games by the loser of the point
 * @private
 */
const calculateLoserGames = (newWinnerGames: number, loserGames:number): number => (
  newWinnerGames === 0 ? 0 : loserGames
);

/**
 * Calculates if the point finishes the current set
 * @param {number} newWinnerGames Won games by the winner (including the current point)
 * @return {boolean} True if as a side effect of winning the point the current set ends
 * @private
 */
const calculateIsFinishingSet = (newWinnerGames: number): boolean => newWinnerGames === 0;

/**
 * Calculates if the match enters in tie break mode
 * @param {number} newWinnerGames Won games by the winner (including the current point)
 * @param {number} loserGames Won games by the loser
 * @return {boolean} True if the match enters in tie break mode, False otherwise
 * @private
 */
const calculateEnterTiebreakMode = (newWinnerGames: number, loserGames: number): boolean =>
  newWinnerGames === MINIMUM_GAMES_TO_WIN_A_SET && loserGames === MINIMUM_GAMES_TO_WIN_A_SET;

/**
  * Calculates the implications of win the point in the match score
  * @param {ScoreContextType} scoreContext The latest point context
  * @returns {NextGameSideEffectType} The side effects of the latest point in the match score
  * @public
  */
const calculateNextGame = ({ winnerName, loserName, matchScore }: ScoreContextType): NextGameSideEffectType => {
  const winnerGames = matchScore[winnerName].games;
  const loserGames = matchScore[loserName].games;

  const newWinnerGames = calculateWinnerGames(winnerGames, loserGames);
  const newLoserGames = calculateLoserGames(newWinnerGames, loserGames);
  const isFinishingSet = calculateIsFinishingSet(newWinnerGames);

  const newWinnerHistoricScoreboard = Array.from(matchScore[winnerName].historicScoreboard);
  const newLoserHistoricScoreboard = Array.from(matchScore[loserName].historicScoreboard);

  if (isFinishingSet) {
    newWinnerHistoricScoreboard.push(newWinnerGames || winnerGames + 1);
    newLoserHistoricScoreboard.push(loserGames);
  }

  const isInTiebreakMode = calculateEnterTiebreakMode(newWinnerGames, loserGames);

  return {
    isFinishingSet,
    updates: [
      {
        keyPath: `${winnerName},historicScoreboard`,
        value: newWinnerHistoricScoreboard,
      },
      {
        keyPath: `${loserName},historicScoreboard`,
        value: newLoserHistoricScoreboard,
      },
      {
        keyPath: 'inTiebreakMode',
        value: isInTiebreakMode,
      },
      {
        keyPath: `${winnerName},games`,
        value: newWinnerGames,
      },
      {
        keyPath: `${loserName},games`,
        value: newLoserGames,
      },
    ],
  };
};

export default calculateNextGame;
