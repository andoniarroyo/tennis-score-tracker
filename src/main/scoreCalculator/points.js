// @flow
import type { ScoreContextType, UpdateSideEffectType } from '../matchScoreCalculator';

export type NextPointSideEffectType = {
    isFinishingGame: boolean,
    updates: Array<UpdateSideEffectType>,
}

/**
 * Gives the new points value from 0
 * @returns {string} The next points value
 * @private
 */
const from0 = () => '15';

/**
 * Gives the new points value from 15
 * @returns {string} The next points value
 * @private
 */
const from15 = () => '30';

/**
 * Gives the new points value from 30
 * @returns {string} The next points value
 * @private
 */
const from30 = () => '40';

/**
 * Gives the new points value from 40, taking into consideration the another
 * player points value
 * @param {string} loserPoints The points of the loser player
 * @returns {string} The next points value
 * @private
 */
const from40 = (loserPoints: string): string => {
  switch (loserPoints) {
    case 'A':
      return '40';
    case '40':
      return 'A';
    default:
      return '0';
  }
};

/**
 * Gives the new points value from advantage
 * @returns {string} The next points value
 * @private
 */
const fromA = () => '0';

/**
 * The set of rules to be applied based on the current points value
 */
const advanceRules = {
  from0,
  from15,
  from30,
  from40,
  fromA,
};

/**
 * Calculates the new winner points value
 * @param {string} winnerScore The winner points value
 * @param {string} loserScore  The loser points value
 * @returns {string} The new winner points value
 * @private
 */
const calculateNewWinerScore = (winnerScore: string, loserScore:string): string =>
  advanceRules[`from${winnerScore}`](loserScore);

/**
 * Calculates the new loser points value
 * @param {string} newWinnerScore The new winner points value
 * @param {string} oldWinnerScore The old winner points value
 * @param {string} loserScore The loser points value
 * @returns {string} The new winner points value
 * @private
 */
const calculateNewLoserScore = (newWinnerScore: string, oldWinnerScore: string, loserScore: string): string => {
  if (newWinnerScore === '0') {
    return '0';
  }
  return newWinnerScore === oldWinnerScore ? '40' : loserScore;
};

/**
 *  Calculates if the point finishes the current game
 * @param {string} newWinnerScore The new winner points score
 * @returns {boolean} True if the point finishes the game, False otherwise
 * @private
 */
const calculateIsFinishingGame = (newWinnerScore: string): boolean => newWinnerScore === '0';

/**
 * Calculates the next points value based on the winner of the current point
 * @param {ScoreContextType} scoreContext score context
 * @returns {NextPointSideEffectType} The next point side effect generated by this point
 * @public
 */
const calculateNextScore = ({ winnerName, loserName, matchScore }: ScoreContextType): NextPointSideEffectType => {
  const winnerScore = matchScore[winnerName].score;
  const loserScore = matchScore[loserName].score;
  const newWinnerScore = calculateNewWinerScore(winnerScore, loserScore);
  const newLoserScore = calculateNewLoserScore(newWinnerScore, winnerScore, loserScore);
  const isFinishingGame = calculateIsFinishingGame(newWinnerScore);

  return {
    isFinishingGame,
    updates: [
      {
        keyPath: `${winnerName},score`,
        value: newWinnerScore,
      },
      {
        keyPath: `${loserName},score`,
        value: newLoserScore,
      },
    ],
  };
};

export default calculateNextScore;
