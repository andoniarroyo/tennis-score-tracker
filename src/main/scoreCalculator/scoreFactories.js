// @flow
// internal types
import type { PlayerScoreType, MatchScoreType } from '../index';
//internal modules
import { MATCH_STATUS_INPLAY } from '../constants';

/**
 * Composes a playscore with the values received by parameters
 * @param {number} sets The number of sets won by the player
 * @param {number} games The number of games won by the player
 * @param {string} score The number of points won by the player
 * @param {number} tieBreak The number of tiebreak points won by the player
 * @param {Array<number>} historicScoreboard The historic scoreboards for the player
 * @returns {PlayerScoreType} The player score generated object
 */
export const playerScoreFactory = (sets: number = 0, games: number = 0, score: string = '0',
  tieBreak: number = 0, historicScoreboard: Array<number> = []): PlayerScoreType => (
  {
    sets, games, score, tieBreak, historicScoreboard,
  }
);

/**
 * Composes a match score based on the received parameters
 * @param {PlayerScoreType} player1Score The player 1 score
 * @param {PlayerScoreType} player2Score The player 2 score
 * @param {boolean} inTiebreakMode The initial value for the in tiebreak mode
 * @returns {MatchScoreType} The match score generated object
 */
export const matchScoreFactory = (player1Score: PlayerScoreType = playerScoreFactory(),
  player2Score: PlayerScoreType = playerScoreFactory(), inTiebreakMode: boolean = false): MatchScoreType => (
  {
    status: MATCH_STATUS_INPLAY,
    winner: '',
    player1: player1Score,
    player2: player2Score,
    inTiebreakMode,
  }
);
