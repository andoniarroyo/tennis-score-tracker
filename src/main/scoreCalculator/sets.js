// @flow
// internal types
import type { ScoreContextType, UpdateSideEffectsType } from '../matchScoreCalculator';

// internal modules
import { SETS_TO_WIN, MATCH_STATUS_INPLAY, MATCH_STATUS_FINISHED } from '../constants';

/**
 * Calculates the new sets value
 * @param {number} winnerSets Won sets by the winner
 * @returns {number} The new won sets by the winner
 * @private
 */
const calculateNewWinnerSets = (winnerSets: number): number => winnerSets + 1;

/**
 * Calculates the match new status
 * @param {number} newWinnerSets Won sets by the winner (including the current point)
 * @returns {string} The new status of the match
 * @private
 */
const calculateMatchStatus = (newWinnerSets: number): string => (
  newWinnerSets === SETS_TO_WIN ? MATCH_STATUS_FINISHED : MATCH_STATUS_INPLAY
);

/**
 * Calculates the winner name of the match if there is one
 * @param {string} matchStatus The current match status
 * @param {string} winner The point winner name
 * @returns {string} The name of the winner if the match is over, empty string otherwise
 * @private
 */
const calculateNewWinner = (matchStatus: string, winner: string) => (
  matchStatus === MATCH_STATUS_FINISHED ? winner : ''
);

/**
  * Calculates the implications of win the set in the match score
  * @param {ScoreContextType} scoreContext The latest point context
  * @returns {UpdateSideEffectsType} The side effects of the latest point in the match score
  * @public
  */
const calculateNextSet = ({ winnerName, matchScore }: ScoreContextType): UpdateSideEffectsType => {
  const winnerSets = matchScore[winnerName].sets;
  const newWinnerSets = calculateNewWinnerSets(winnerSets);
  const matchStatus = calculateMatchStatus(newWinnerSets);
  const matchWinner = calculateNewWinner(matchStatus, winnerName);

  return [
    {
      keyPath: `${winnerName},sets`,
      value: newWinnerSets,
    },
    {
      keyPath: 'status',
      value: matchStatus,
    },
    {
      keyPath: 'winner',
      value: matchWinner,
    },
  ];
};

export default calculateNextSet;
