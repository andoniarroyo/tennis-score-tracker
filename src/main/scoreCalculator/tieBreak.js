// @flow
// internal types
import type { ScoreContextType, UpdateSideEffectsType } from '../matchScoreCalculator';
// internal modules
import { SETS_TO_WIN, MATCH_STATUS_FINISHED } from '../constants';

const MINIMUM_POINTS_TO_WIN_A_TIEBREAK = 7;
const MINIMUM_GAMES_DIFFERENCE_TO_WIN_A_TIBREAK = 2;

/**
 * Calculates if the tiebreak is finished bacuase of the finished point
 * @param {number} winnerTiebreak The winner tiebreak value
 * @param {number} loserTiebreak The loser tiebreak value
 * @returns {boolean} True if the tiebreak is finished, false otherwise
 * @private
 */
const calculateIsFinishingTiebreak = (winnerTiebreak: number, loserTiebreak: number): boolean => (
  winnerTiebreak >= MINIMUM_POINTS_TO_WIN_A_TIEBREAK &&
    (winnerTiebreak - loserTiebreak) >= MINIMUM_GAMES_DIFFERENCE_TO_WIN_A_TIBREAK
);

/**
* Calculates the implications of win a point in the tiebreal for the match score
* @param {ScoreContextType} scoreContext The latest point context
* @returns {UpdateSideEffectType} The side effects of the latest point in the match score
* @public
 */
const calculateNextTieBreak = ({ winnerName, loserName, matchScore }: ScoreContextType): UpdateSideEffectsType => {
  let winnerTieBreak = matchScore[winnerName].tieBreak + 1;
  let loserTieBreak = matchScore[loserName].tieBreak;
  let winnerGames = matchScore[winnerName].games;
  let loserGames = matchScore[loserName].games;
  let winnerSets = matchScore[winnerName].sets;
  let { status } = matchScore;
  let matchWinner = '';

  const newWinnerHistoricScoreboard = Array.from(matchScore[winnerName].historicScoreboard);
  const newLoserHistoricScoreboard = Array.from(matchScore[loserName].historicScoreboard);

  const isFinishingTiebreak = calculateIsFinishingTiebreak(winnerTieBreak, loserTieBreak);

  if (isFinishingTiebreak) {
    newWinnerHistoricScoreboard.push(winnerGames + 1);
    newLoserHistoricScoreboard.push(loserGames);
    winnerGames = 0;
    loserGames = 0;
    winnerSets = matchScore[winnerName].sets + 1;
    winnerTieBreak = 0;
    loserTieBreak = 0;
    if (winnerSets === SETS_TO_WIN) {
      status = MATCH_STATUS_FINISHED;
      matchWinner = winnerName;
    }
  }

  return [
    {
      keyPath: `${winnerName},historicScoreboard`,
      value: newWinnerHistoricScoreboard,
    },
    {
      keyPath: `${loserName},historicScoreboard`,
      value: newLoserHistoricScoreboard,
    },
    {
      keyPath: 'inTiebreakMode',
      value: !isFinishingTiebreak,
    },
    {
      keyPath: `${winnerName},tieBreak`,
      value: winnerTieBreak,
    },
    {
      keyPath: `${loserName},tieBreak`,
      value: loserTieBreak,
    },
    {
      keyPath: `${winnerName},games`,
      value: winnerGames,
    },
    {
      keyPath: `${loserName},games`,
      value: loserGames,
    },
    {
      keyPath: `${winnerName},sets`,
      value: winnerSets,
    },
    {
      keyPath: 'status',
      value: status,
    },
    {
      keyPath: 'winner',
      value: matchWinner,
    },
  ];
};

export default calculateNextTieBreak;
