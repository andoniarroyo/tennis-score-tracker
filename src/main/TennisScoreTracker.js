// @flow
// external modules
import React from 'react';
// internal types
import type { MatchScoreType } from './index';
import type { UpdateScoreActionType } from './actions';
// internal modules
import { matchScoreFactory } from './scoreCalculator/scoreFactories';
import HistoricScoreboard from './components/HistoricScoreboard';
import MatchScoreTracker from './components/MatchScoreTracker';

import './styles/tennisScoreTracker.scss';

export type TennisScoreTrackerPropsType = {
    score: MatchScoreType,
    updateScore: (matchScore: MatchScoreType) => UpdateScoreActionType
};

/**
 * The tennis score tracker render method
 * @param {TennisScoreTrackerPropsType} The properties to render the object
 * @returns {React$Element<any>} HTML markup for the component
 */
const TennisScoreTracker = ({ score, updateScore }: TennisScoreTrackerPropsType) => (
  <div>
    <div className="title-section">
      <div>Status: {score.status}</div>
      <div>Winner: {score.winner || '...'}</div>
    </div>
    <MatchScoreTracker score={score} updateScore={updateScore} />
    <HistoricScoreboard player1historicScoreboard={score.player1.historicScoreboard}
      player2historicScoreboard={score.player2.historicScoreboard}
    />
    <div className="title-section">
      <div>
        Other actions:
      </div>
      <button onClick={() => updateScore(matchScoreFactory())}>
        Reset the match and start again!
      </button>
    </div>
  </div>
);

export default TennisScoreTracker;
