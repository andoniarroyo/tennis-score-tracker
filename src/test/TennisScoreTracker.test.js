import React from 'react';
import renderer from 'react-test-renderer';

import TennisScoreTracker from '../main/TennisScoreTracker';
import { matchScoreFactory } from '../main/scoreCalculator/scoreFactories';

describe('<TennisScoreTracker />', () => {
  it('renders the tennis score tracker', () => {
    const props = {
      score: matchScoreFactory(),
      updateScore: () => {},
    };

    const tree = renderer.create(<TennisScoreTracker {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
