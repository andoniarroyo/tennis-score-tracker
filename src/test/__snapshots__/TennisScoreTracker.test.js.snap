// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TennisScoreTracker /> renders the tennis score tracker 1`] = `
<div>
  <div
    className="title-section"
  >
    <div>
      Status: 
      inplay
    </div>
    <div>
      Winner: 
      ...
    </div>
  </div>
  <div
    className="container"
  >
    <div>
      <div
        className="player-name"
      >
        player1
      </div>
      <div
        className="player-score-content"
      >
        <div>
          sets:
        </div>
        <div>
          0
        </div>
        <div>
          games:
        </div>
        <div>
          0
        </div>
        <div>
          score:
        </div>
        <div
          className="score-main-value"
        >
          0
        </div>
        <div
          className="scored"
        >
          <button
            disabled={false}
            onClick={[Function]}
          >
            player1
             scored!
          </button>
        </div>
      </div>
    </div>
    <div
      className="tiebreak-mode-false"
    >
      <div>
        Tiebreak:
      </div>
      <div
        className="tiebreak-content"
      >
        <div
          className="score-tiebreak"
        >
          0
        </div>
        <div
          className="score-tiebreak"
        >
          0
        </div>
      </div>
    </div>
    <div>
      <div
        className="player-name"
      >
        player2
      </div>
      <div
        className="player-score-content"
      >
        <div>
          sets:
        </div>
        <div>
          0
        </div>
        <div>
          games:
        </div>
        <div>
          0
        </div>
        <div>
          score:
        </div>
        <div
          className="score-main-value"
        >
          0
        </div>
        <div
          className="scored"
        >
          <button
            disabled={false}
            onClick={[Function]}
          >
            player2
             scored!
          </button>
        </div>
      </div>
    </div>
  </div>
  <table>
    <thead>
      <tr>
        <td>
          <span>
            Scoreboard:
          </span>
        </td>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          className="player-historic-name"
        >
          player 1:
        </td>
      </tr>
      <tr>
        <td
          className="player-historic-name"
        >
          player 2:
        </td>
      </tr>
    </tbody>
  </table>
  <div
    className="title-section"
  >
    <div>
      Other actions:
    </div>
    <button
      onClick={[Function]}
    >
      Reset the match and start again!
    </button>
  </div>
</div>
`;
